crates/table/src/read_column.rs

```diff
@@ -0,0 +1,507 @@
+        ///
+        /// That is, test that, for any row type and any row value,
+        /// inserting the row, then doing `AlgebraicValue::read_column` on each column of the row
+        /// returns the expected value.
+        fn read_column_same_value((ty, val) in generate_typed_row()) {
+            let mut blob_store = HashMapBlobStore::default();
+            let mut table = table(ty);
+
+            let (_, ptr) = table.insert(&mut blob_store, &val).unwrap();
+
+            let row_ref = table.get_row_ref(&blob_store, ptr).unwrap();
+
+            for (idx, orig_col_value) in val.elements.into_iter().enumerate() {
+                let read_col_value = AlgebraicValue::read_column(row_ref, idx).unwrap();
+                prop_assert_eq!(orig_col_value, read_col_value);
+            }
+        }
+
+        #[test]
+        /// Test that trying to read a column at a type more specific than `AlgebraicValue`
+        /// which does not match the actual column type
+        /// returns an appropriate error.
+        fn read_column_wrong_type((ty, val) in generate_typed_row()) {
+            let mut blob_store = HashMapBlobStore::default();
+            let mut table = table(ty.clone());
+
+            let (_, ptr) = table.insert(&mut blob_store, &val).unwrap();
+
+            let row_ref = table.get_row_ref(&blob_store, ptr).unwrap();
+
+            for (idx, col_ty) in ty.elements.into_iter().enumerate() {
+                assert_wrong_type_error::<u8>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::U8)?;
+                assert_wrong_type_error::<i8>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::I8)?;
+                assert_wrong_type_error::<u16>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::U16)?;
+                assert_wrong_type_error::<i16>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::I16)?;
+                assert_wrong_type_error::<u32>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::U32)?;
+                assert_wrong_type_error::<i32>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::I32)?;
+                assert_wrong_type_error::<u64>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::U64)?;
+                assert_wrong_type_error::<i64>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::I64)?;
+                assert_wrong_type_error::<u128>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::U128)?;
+                assert_wrong_type_error::<i128>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::I128)?;
+                assert_wrong_type_error::<f32>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::F32)?;
+                assert_wrong_type_error::<f64>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::F64)?;
+                assert_wrong_type_error::<bool>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::Bool)?;
+                assert_wrong_type_error::<String>(row_ref, idx, &col_ty.algebraic_type, AlgebraicType::String)?;
```

Comment:
that's an entertaining pattern

Reply:
Can you tell I miss Common Lisp having types at runtime?